### DISCUSSION TITLE
Add/Edit Geometries in mayo

### DISCUSSION THREAD

**Discussion by tjumsl** [fougue/mayo#81] [Ideas]
First! Thanks for developing the wonderful cad viewer mayo, it it really a very good software!

If possible, coudld you please kindly help to provide some details regarding how geometries can be added / edited in mayo using occ interfaces? This would be helpful for some simple geometry creatations and operations (such as boolen etc).

**Comment by HuguesDelorme** [comment]
Thanks for the compliment, that's appreciated!
Concerning your question, yes it's 100% possible to add this kind of features for shapes creation.
It should be fairly easy :

1) Create a new class that inherits from `Mayo::Command`. This base class is declared in `src/app/commands_api.h`
   To implement this new class(called for example `MyCommandCreateBox`) have a look at the code of some existing commands(eg `CommandMainWidgetToggleFullscreen` in `src/app/commands_window.h`

2) Add your new command in function `MainWindow::createCommands()` in file `src/app/mainwindow.cpp`
    Example:
````cpp
void MainWindow::createCommands()
{
    // ...
    this->addCommand<MyCommandCreateBox>();
}
````

 3) Insert the command into some existing menu item in function `MainWindow::createMenus()`
     Example:

````cpp
void MainWindow::createMenus()
{
    // ...
    {   // Tools
        auto menu = m_ui->menu_Tools;
        fnAddAction(menu, CommandSaveViewImage::Name);
        fnAddAction(menu, CommandInspectXde::Name);
        menu->addSeparator();
        fnAddAction(menu, CommandEditOptions::Name);
        menu->addSeparator();
        fnAddAction(menu, MyCommandCreateBox::Name); // <-- YOUR COMMAND ACTION INSERTED HERE
    }
````

4) Implement function `MyCommandCreateBox::execute()`
    Example:
````cpp
void MyCommandCreateBox::execute()
{
    auto guiDoc = this->context()->currentGuiDocument();
    auto doc = guiDoc ? guiDoc->document() : nullptr;
    if (doc) {
        // Create box BREP shape
        const TopoDS_Shape shapeBox = BRepPrimAPI_MakeBox(100, 100, 100).Solid();
        // Declare a label in the XDE structure to store the created shape
        const TDF_Label labelShape = doc->xcaf().shapeTool()->NewShape();
        doc->xcaf().shapeTool()->SetShape(labelShape, shapeBox);
        // Set a name associated to the shape
        TDataStd_Name::Set(labelShape, "MyBox");
        // Add root entity to the document, throws signal Document::signalEntityAdded caught by GuiApplication that
        // will automatically create/add 3D graphics in the corresponding GuiDocument object
        doc->addEntityTreeNode(labelShape); 
    }
}
````

**Reply by boredumbk** [reply]
@HuguesDelorme Hello, I have used the above code snippets to add add geometries but I want to trigger the generation of the geometry on the mouse click and on the point where the cursor is clicked. I edited widget_occ_view_controller.cpp to have a sphere generation function on mouse click when geometry adding mode is "ON" (I have created a geometry add mode in the menus create function), but I am having issues in detecting the currentGuiDocument in widget_occ_view_controller. Any other way to tackle this issue or any other alternative way of producing additional geometry on mouse clicks when a mode from the menu is toggled "ON"?

**Reply by HuguesDelorme** [reply]
@boredumbk 
Hello, you shouldn't implement your logic within `WidgetOccViewController` but rather in your new `Command` sub class(let's call it `MyCommand`)
Basically in `MyCommand::execute()` member function you should add this signal/slot connection:
````c++
if (this->action()->isChecked()) {
    auto viewCtrl = this->context()->v3dViewController(this->currentGuiDocument());
    viewCtrl->signalMouseButtonClicked.connectSlot([=](Aspect_VKeyMouse btn) {
        if (btn == Aspect_VKeyMouse_RightButton) {
            // Code to create the sphere...
        }
    });
}
````
This code doesn't deal with multi documents handling(users can open multiple files in Mayo in the same session)
`MyCommand` will have to keep the `SignalConnectionHandle` object returned by `connectSlot()` and disconnect it when the command is toggled off.

**Reply by boredumbk** [reply]
Thanks to your previous response @HuguesDelorme, I am able to generate additional geometry based on right click, but I am not able to generate the geometry on the clicked location as signalMouseButtonClicked doesn't provide the coordinates of the clicked location, how to go about getting the coordinates of the clicked location? Further if I had to produce geometries on left click when defects mode is turned on do I have to deactivate the standard function mapped with left click that is selecting parts of the assembly of the rendered model?

**Reply by HuguesDelorme** [reply]
@boredumbk
Mouse location is indeed not conveyed with `V3dViewController::signalMouseButtonClicked()`
To workaround this you can also connect to signal `V3dViewController::signalMouseMoved()` add track what's emitted

I'm not sure to understand your last question

**Reply by boredumbk** [reply]
Sorry I should have framed it better.
I had another question, I added another signal to get the mouse location when the right mouse button is clicked but while generating the sphere I am facing issues. I want the spheres to be generated on the rendered 3d model but I'm getting it all over the place. Is there any way to get the coordinates of the 3d model which is under the mouse click?
![Screenshot (605)](https://github.com/fougue/mayo/assets/79220849/791b990d-f5f5-42ac-a190-c6419bba07a2)

**Reply by HuguesDelorme** [reply]
@boredumbk
In your slot connected to `signalMouseButtonClicked` you can get the "picked point" like this:
````c++
auto selector = guiDoc->graphicsScene()->mainSelector();
selector->Pick(xMousePos, yMousePos, guiDoc->v3dView());
const gp_Pnt pos3d =
        selector->NbPicked() > 0 ?
                selector->PickedPoint(1) :
                GraphicsUtils::V3dView_to3dPosition(guiDoc->v3dView(), xMousePos, yMousePos);
````

**Reply by boredumbk** [reply]
This is my code for generating the sphere after applying the modifications that you proposed the error has reduced but the spheres are still not on the rendered 3D model, is there anything that I'm doing wrong or are there any ways to resolve this issue 
`void CommandGenerateNode::execute() {
    qDebug("Regrets");

    auto guiDoc = this->currentGuiDocument();
    if (!guiDoc) return; // Early exit if no GUI document is available

    auto viewCtrl = this->context()->v3dViewController(guiDoc);
    if (!viewCtrl) return; // Early exit if no view controller is available

    viewCtrl->signalMouseRightButtonClicked.connectSlot([=](int x, int y) {
        auto selector = guiDoc->graphicsScene()->mainSelector();
        selector->Pick(x, y, guiDoc->v3dView());
        const gp_Pnt pos3d = selector->NbPicked() > 0 ?
            selector->PickedPoint(1) :
            GraphicsUtils::V3dView_to3dPosition(guiDoc->v3dView(), x, y);

        auto doc = guiDoc->document();
        if (doc) {
            const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 10.0).Solid();
            gp_Trsf aTrsf;
            aTrsf.SetTranslation(pos3d.Coord()); // Corrected variable name
            BRepBuilderAPI_Transform aTransformer(shapeSphere, aTrsf);
            TopoDS_Shape transformedSphere = aTransformer.Shape();

            const TDF_Label labelShape = doc->xcaf().shapeTool()->NewShape();
            doc->xcaf().shapeTool()->SetShape(labelShape, transformedSphere);
            TDataStd_Name::Set(labelShape, "MySphere");
            doc->addEntityTreeNode(labelShape);

            qDebug("Sphere generated at mouse click position: [%d, %d]", x, y);
        }
        });
}` 
![Screenshot (606)](https://github.com/fougue/mayo/assets/79220849/a1f6512f-eb84-480c-adc7-86cd73d2735f)

**Reply by HuguesDelorme** [reply]
@boredumbk 
You don't need the gp_Trsf code as you already create the sphere shape at the right location, should be:
````c++
if (doc) {
    const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 10.0).Solid();
    const TDF_Label labelShape = doc->xcaf().shapeTool()->NewShape();
    // ...    
````

**Reply by boredumbk** [reply]
The error has significantly reduced but still it's not accurate 
![Screenshot (607)](https://github.com/fougue/mayo/assets/79220849/967deee4-adb7-4034-963b-c79bc57a9f4e)
When I clicked at the point highlighted by blue circle the sphere generated at the point highlighted by red circle. Is there a way to increase the accuracy?

**Reply by HuguesDelorme** [reply]
@boredumbk
You should check(by debug trace) the computed `pos3d` is the same as the 3D coords displayed in the top right corner of Mayo's mainwindow:
![image](https://github.com/fougue/mayo/assets/939132/e67c9067-ad8f-43ec-a0a3-e329e5fee3b0)

**Reply by boredumbk** [reply]
They are the same 
![Screenshot (611)](https://github.com/fougue/mayo/assets/79220849/0633190b-368d-4190-963f-3c75c8c02493)
![Screenshot (610)](https://github.com/fougue/mayo/assets/79220849/a32643d9-bf5e-42a3-80cf-af9c6cfa515e)
(Please ignore the last debug statement sphere generated at mouse click position cause it displays the location of mouse click received from `signalMouseRightButtonClicked.connectSlot([=](int x, int y)` I'm generating the sphere at the pos3d coordinates only `const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 1.0).Solid();`)

**Reply by HuguesDelorme** [reply]
@boredumbk
Works fine for me:
![spheres-under-mouse](https://github.com/fougue/mayo/assets/939132/0e0b972e-6489-44c2-a5cc-421e820b2a78)
Here is what I tried in `src/app/widget_main_control.cpp` circa L370:
````c++
    // React to mouse move in 3D view:
    //   * update highlighting
    //   * compute and display 3D mouse coordinates(by silent picking)
    static int gmouseX, gmouseY; // Just for testing
    widgetCtrl->signalMouseMoved.connectSlot([=](int xPos, int yPos) {
        const double dpRatio = this->devicePixelRatioF();
        gfxScene->highlightAt(xPos * dpRatio, yPos * dpRatio, guiDoc->v3dView());
        widget->view()->redraw();
        auto selector = gfxScene->mainSelector();
        selector->Pick(xPos, yPos, guiDoc->v3dView());
        const gp_Pnt pos3d =
                selector->NbPicked() > 0 ?
                    selector->PickedPoint(1) :
                    GraphicsUtils::V3dView_to3dPosition(guiDoc->v3dView(), xPos, yPos);
        gmouseX = xPos; // New code
        gmouseY = yPos; // New code
        m_ui->label_ValuePosX->setText(QString::number(pos3d.X(), 'f', 3));
        m_ui->label_ValuePosY->setText(QString::number(pos3d.Y(), 'f', 3));
        m_ui->label_ValuePosZ->setText(QString::number(pos3d.Z(), 'f', 3));
    });

    // New code:
    widgetCtrl->signalMouseButtonClicked.connectSlot([=](Aspect_VKeyMouse mouseBtn) {
        if (mouseBtn == Aspect_VKeyMouse_MiddleButton) {
            auto doc = guiDoc->document();
            auto selector = gfxScene->mainSelector();
            selector->Pick(gmouseX, gmouseY, guiDoc->v3dView());
            if (selector->NbPicked() > 0) {
                const gp_Pnt pos3d = selector->PickedPoint(1);
                const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 1.5).Solid();
                const TDF_Label labelShape = doc->xcaf().shapeTool()->NewShape();
                doc->xcaf().shapeTool()->SetShape(labelShape, shapeSphere);
                TDataStd_Name::Set(labelShape, "MySphere");
                doc->addEntityTreeNode(labelShape);
            }
        }
    });
````

**Reply by boredumbk** [reply]
The issue I'm facing is with every mouse click the scene zooms out and hence the coordinates change in the log, is there a way to disable the custom mouse events like zooming, panning on mouse clicks when I'm generating the spheres       
 
    `viewCtrl->signalMouseRightButtonClicked.connectSlot([=](int` x, int y) {
        qDebug() << x << " " << y;
        auto selector = guiDoc->graphicsScene()->mainSelector();
        selector->Pick(x, y, guiDoc->v3dView());
        const gp_Pnt pos3d = selector->NbPicked() > 0 ?
            selector->PickedPoint(1) :
            GraphicsUtils::V3dView_to3dPosition(guiDoc->v3dView(), x, y);

        qDebug() << "pos3d Coordinates: X=" << pos3d.X() << " Y=" << pos3d.Y() << " Z=" << pos3d.Z();

        auto doc = guiDoc->document();
        if (doc) {

            const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 1.0).Solid();
            qDebug() << "pos3d Coordinates: X=" << pos3d.X() << " Y=" << pos3d.Y() << " Z=" << pos3d.Z();
            qDebug() << x << " " << y;
            const TDF_Label labelShape = doc->xcaf().shapeTool()->NewShape();
            doc->xcaf().shapeTool()->SetShape(labelShape, shapeSphere);

            TDataStd_Name::Set(labelShape, "MySphere");

            doc->addEntityTreeNode(labelShape);

            //qDebug()<<"Sphere generated at mouse click position:";
        }
        });`
![Screenshot (615)](https://github.com/fougue/mayo/assets/79220849/efa9f616-43df-48a7-836c-7c631ebeb6a8)


These are the spheres generated while clicking at the same position

**Reply by HuguesDelorme** [reply]
@boredumbk
Try to comment this line https://github.com/fougue/mayo/blob/8e3302fe93d7092913f2e061bc32866b26be3501/src/gui/gui_document.cpp#L644
and check if it solves your issue

**Reply by boredumbk** [reply]
It doesn't resolve the issue, after commenting the line you suggested, if I don't rotate the 3D object it appears that the spheres are getting generated at the same location but the moment I rotate it, I can see the spheres piling on each other 
![Screenshot (618)](https://github.com/fougue/mayo/assets/79220849/6c644952-9c3d-4e0d-934b-2b37d5ca4b96)
This is how it looks and it appears that the spheres are getting generated at the same location but the moment I rotate it 
![Screenshot (619)](https://github.com/fougue/mayo/assets/79220849/882e5371-cb52-49a9-adbd-e9d3ce804abd)

**Reply by HuguesDelorme** [reply]
I guess this is just that the next pick is hitting the latest created sphere shape hence the "stack" you get
You will get better behavior by adding the "sphere" graphics yourself and explicitly remove it from the list of "pickable" objects:
````c++
    // React to mouse move in 3D view:
    //   * update highlighting
    //   * compute and display 3D mouse coordinates(by silent picking)
    static int gmouseX, gmouseY; // Just for testing
    widgetCtrl->signalMouseMoved.connectSlot([=](int xPos, int yPos) {
        const double dpRatio = this->devicePixelRatioF();
        gfxScene->highlightAt(xPos * dpRatio, yPos * dpRatio, guiDoc->v3dView());
        widget->view()->redraw();
        auto selector = gfxScene->mainSelector();
        selector->Pick(xPos, yPos, guiDoc->v3dView());
        const gp_Pnt pos3d =
            selector->NbPicked() > 0 ?
                selector->PickedPoint(1) :
                GraphicsUtils::V3dView_to3dPosition(guiDoc->v3dView(), xPos, yPos);
        gmouseX = xPos; // New code
        gmouseY = yPos; // New code
        m_ui->label_ValuePosX->setText(QString::number(pos3d.X(), 'f', 3));
        m_ui->label_ValuePosY->setText(QString::number(pos3d.Y(), 'f', 3));
        m_ui->label_ValuePosZ->setText(QString::number(pos3d.Z(), 'f', 3));
    });

    // New code:
    widgetCtrl->signalMouseButtonClicked.connectSlot([=](Aspect_VKeyMouse mouseBtn) {
        if (mouseBtn == Aspect_VKeyMouse_MiddleButton) {
            auto selector = gfxScene->mainSelector();
            selector->Pick(gmouseX, gmouseY, guiDoc->v3dView());
            if (selector->NbPicked() > 0) {
                const gp_Pnt pos3d = selector->PickedPoint(1);
                const TopoDS_Shape shapeSphere = BRepPrimAPI_MakeSphere(pos3d, 1.5).Solid();
                Handle_AIS_Shape aisSphere = new AIS_Shape(shapeSphere);
                aisSphere->SetDisplayMode(AIS_Shaded);
                gfxScene->addObject(aisSphere);
                gfxScene->redraw();
                selector->RemoveSelectableObject(aisSphere);
            }
        }
    });
````

**Reply by itsivag** [reply]
I have also tried replicating the same thing here's what I got, the coordinate of mouse click, and the sphere generation is different.


https://github.com/fougue/mayo/assets/85468732/bdf00465-18e3-419c-be7d-8da4a3444780

**Reply by itsivag** [reply]
Any updates on this query?

**Reply by HuguesDelorme** [reply]
@itsivag
Try to check the mouse position(`currPos` variable) is correct within this function:https://github.com/fougue/mayo/blob/7445a2e9e3325b26f6fc735e3005aef2d721e692/src/app/widget_occ_view_controller.cpp#L257
What is your OS and OpenCascade version? Please past here the first 20 lines of `Help->System Information...` command output

**Reply by itsivag** [reply]
```
{noformat}
Qt 6.6.0 (x86_64-little_endian-llp64 shared (dynamic) release build; by MSVC 2019) on "windows" 
OS: Windows 11 Version 22H2 [winnt version 10.0.22621]

Architecture: x86_64; features: SSE2 SSE3 SSSE3 SSE4.1 SSE4.2 AVX AVX2 RDRAND RDSEED AES SHA

Environment:
  QTDIR="C:\OpenCASCADE-7.5.0-vc14-64\opencascade-7.5.0\..\qt5.11.2-vc14-64"
  QtMsBuild="C:\Users\itsivag\AppData\Local\QtMsBuild"
  QUARTUS_ROOTDIR="C:\altera\13.1\quartus"

Features: QT_NO_EXCEPTIONS

Library info:
  PrefixPath: C:\Qt\Tools\QtCreator\bin
  DocumentationPath: C:\Qt\Tools\QtCreator\bin\doc
  HeadersPath: C:\Qt\Tools\QtCreator\bin\include
  LibrariesPath: C:\Qt\Tools\QtCreator\bin
  LibraryExecutablesPath: C:\Qt\Tools\QtCreator\bin\bin
  BinariesPath: C:\Qt\Tools\QtCreator\bin
  PluginsPath: C:\Qt\Tools\QtCreator\bin\plugins
  QmlImportsPath: C:\Qt\Tools\QtCreator\bin\qml
  ArchDataPath: C:\Qt\Tools\QtCreator\bin
  DataPath: C:\Qt\Tools\QtCreator\bin
  TranslationsPath: C:\Qt\Tools\QtCreator\bin\translations
  ExamplesPath: C:\Qt\Tools\QtCreator\bin\examples
  TestsPath: C:\Qt\Tools\QtCreator\bin\tests
  SettingsPath: 
```

**Reply by itsivag** [reply]
are you on linux?

**Reply by HuguesDelorme** [reply]
@itsivag 
No my OS is Win10
I tried the code to create sphere shapes under the mouse with Qt 6.5.2 and OpenCascade 7.5.0 and it just works fine
Don't know why there's this issue on your side, maybe you have a 4k screen?

**Reply by itsivag** [reply]
Nope I have a 1080p display and i don't think win 11 causes the issue.

**Reply by HuguesDelorme** [reply]
@itsivag
Are you working within GitHub fork of Mayo project? If so you can paste URL here and I will try directly your code

**Reply by itsivag** [reply]
I'll create one and share

**Reply by itsivag** [reply]
i think you can edit this fork https://github.com/itsivag/mayo

**Reply by HuguesDelorme** [reply]
@itsivag
I had a look and it's using the previous version of my code suggestion to @boredumbk (where the sphere shape is "pickable" and leading to unwanted to behavior)
Aside from that it's working as expected, the sphere shape is created and located at the picked position
You will have to investigate, try to put some traces and as I mentionned please check if the mouse coords are correct

**Reply by itsivag** [reply]
yeah, i used the prev one as i needed the spheres to be pickable but your reply though having a pickable sphere it is spawned on the correct mouse coords but that's not the case for me.In this my cursor is no where near to the sphere

![image](https://github.com/fougue/mayo/assets/85468732/b3f017fe-8a6b-4292-ad29-49b8a4efe359)
